### ISSUE TITLE
Chrome applications (including electron) that use system tray sets empty tray icon when private-tmp is active

### ISSUE THREAD

**Issue by msva** [netblue30/firejail#1137]
I've found that `slack` (which have `private-tmp` active in default system-wide profile) started to render empty icon in the tray.
Debugging showed that on every start it creates two directories under `/tmp`, named like `.org.chromium.Chromium.[6 random alphanumeric]`. And one of them contains `icons/hicolor/24x24/apps` path, and in the `apps` directory, it places is a files (may be more than one), that named like: `chrome_app_indicator2_[32 random hex digit].png`.

I guess, it tells systemtray to load icon from that path, but since the path exists only in private bin, system tray fails to do that.

I'd try to work it around with `whitelist /tmp/.org.chromium.Chromium.*`, but... `whitelist` (and friends) does not support regex :)

**Comment by netblue30** [comment]
I would try to comment out (add a #)  private-tmp in /etc/firejail/slack.profile.

**Comment by msva** [comment]
> I would try to comment out (add a #)  private-tmp in
> /etc/firejail/slack.profile.

Yes, it helps (that is how I debugged that it is private-tmp's fault).

But it has side-effects:
1) it will (obviously) expose entire /tmp to slack
2) that change will be rewritten after next firejail upgrade.

**Comment by netblue30** [comment]
2) is easy, I'll remove it from the profile. For 1), can you figure out what files or directories it uses in /tmp? We can  whitelist /tmp. At minimum, we need just a "whitelist /tmp/.X11-unix" and the directories/files created by slack.

**Comment by msva** [comment]
as far as I looked in private /tmp inside jail, it only creates directory 
"Slack Crashes" and two directories I described above 
(".org.chromium.Chromium.[a-zA-Z0-9]{6}"), and uses nothing from system's /tmp

**Comment by netblue30** [comment]
OK, so I only need to add some basic globbing/regx support in whitelists, something like "whitelist .org.chromium.Chromium.*".

**Comment by msva** [comment]
Hi there!

Is it ant progress about this? 😎

**Comment by msva** [comment]
I'm not sure if whitelist now supports wildcards (at least, it doesn't compain on `*`), but this is still the issue.
And, I'm not sure, if it will work as intended out of the box, since it **creates** that directories (containing the tray icon png file) in runtime (i.e. after start), and they have unique names every time.
So, I'm not sure, if it will be enough to just whitelist that directories (with wildcard), or will firejail need to "proxy" that directories in "outter" /tmp?..
How do you think, @netblue30 ?

**Comment by msva** [comment]
By the way, the same thing is actual for chrome forks. For example, Yandex Browser just created `/tmp/.ru.yandex.desktop.browser.MJGCGo` directory. I bet new Opera doing it in the similar way.
So, we definitely need a way to somehow whitelist not-yet-created directories by mask (to not just manually blacklist files from /tmp to prevent proprietary things to access, but just sandbox them in their own /tmp, but having some directories they creating in outter /tmp.

Or, maybe, do a LD_PRELOAD hack to hijack that files creation/writing to them and do the same in "external" copy in "outter" /tmp.

**Comment by chiraag-nataraj** [comment]
Hey all,

I'm actually currently using `private-tmp` with `slack` and it's actually the _only_ way it works (apparently Electron apps need `/tmp` to be executable, and my `/tmp` is mounted `noexec`). I've attached the profile I'm using currently - can someone else test and report back if it works for them?
[slack.txt](https://github.com/netblue30/firejail/files/1860620/slack.txt)

**Comment by alien2003** [comment]
The same with wire-desktop

**Comment by SkewedZeppelin** [comment]
@alien2003 
try adding the following to `~/.config/firejail/slack.local`
```
noblacklist ${HOME}/.local/share/gnome-shell
whitelist ${HOME}/.local/share/gnome-shell
read-only ${HOME}/.local/share/gnome-shell
```

**Comment by alien2003** [comment]
@SkewedZeppelin No errors in terminal now but the icon is still missing. I use Gnome 3.30.2 on Manjaro with this extension for tay icons: https://extensions.gnome.org/extension/615/appindicator-support/

**Comment by alien2003** [comment]
Seems to be ok with this extension: https://extensions.gnome.org/extension/1503/tray-icons/

**Comment by intika** [comment]
Same issue with skype... as its chromium based

**Comment by intika** [comment]
**Ohhh i found a quick easy solution !**

**dbus-lauch** baby :D

just prefix the binary with dbus-lauch...

**Comment by smitsohu** [comment]
Whitelist globbing was added recently by @netblue30, so `whitelist /tmp/.org.chromium.Chromium.*` should work now in Firejail master.
Could someone confirm?

**Comment by rusty-snake** [comment]
@smitsohu if I got this right, `/tmp/foobar.XXXX` is create by foobar. This still did not help.

**Comment by smitsohu** [comment]
@rusty-snake I see. Then the only solutions are indeed
* `ignore private-tmp`
* @intika 's proposal

**Comment by kortewegdevries** [comment]
Is it only for me or the tray icon works without any fix? `yay -S wir-desktop` `firejail-git`. https://imgur.com/HFkR5rb

**Comment by rusty-snake** [comment]
result from #3540:
```
ignore private-tmp

mkdir /tmp/FOO-tmp
whitelist /tmp/FOO-tmp
env TMPDIR=/tmp/FOO-tmp
env TMP=/tmp/FOO-tmp

# IF dbus-user filter
dbus-user.talk org.freedesktop.StatusNotifierItem
dbus-user.talk org.kde.StatusNotifierWatcher
# ELIF dbus-user none
ignore dbus-user none
dbus-user filter
dbus-user.talk org.freedesktop.StatusNotifierItem
dbus-user.talk org.kde.StatusNotifierWatcher
# FI
```

---

Would it make sense to add a `private-tmp-exposed` which does the `/tmp/FOO-tmp` stuff?
